File: erlang/filters/refilter.erl
% Typical req.query (at a minimum, the "filter" parameter will be there):
%
% {[ { <<"filter">>, <<"usagemover/field_regex">> }
%  , { <<"field">> , <<"_id">> }
%  , { <<"regex">> , <<"^a">> }
%  , { <<"log">>   , true } % (optional)
% ]}

fun({Doc}, {Req})
    -> {Query} = couch_util:get_value(<<"query">>, Req)

    % Unfortunately, building a function for every execution (i.e. once per doc) imposes a steep
    % performance penalty. I tried the process dictionary to memoize it but that didn't help.
    % So, uncomment this to activate logging. In the Javascript version, performance is secondary,
    % so it should support logging by default.
    %, Logger = case couch_util:get_value(<<"log">>, Query)
    %    of true      -> fun(F, A) -> Log(lists:flatten(io_lib:format(F, A))) end
    %    ; <<"true">> -> fun(F, A) -> Log(lists:flatten(io_lib:format(F, A))) end
    %    ; _          -> fun(_F, _A) -> ok end
    %    end

    , Field = couch_util:get_value(<<"field">>, Query)
    , Regex = couch_util:get_value(<<"regex">>, Query)
    %, Logger("Running: Id=~p\nReq=~p", [Id, Req])

    , case is_binary(Field) andalso is_binary(Regex)
        of false ->
            %Logger("Invalid parameters: Field=~p Regex=~p", [Field, Regex]),
            exit({invalid_query, <<"Required field and regex parameter">>})
        ; true
            %Logger("Field=~p Regex=~p", [Field, Regex]),
            -> Is_match = case couch_util:get_value(Field, Doc)
                of undefined ->
                    %Logger("Doc ~s has no field ~s", [Id, Field]),
                    false
                ; Non_bin when not is_binary(Non_bin) ->
                    %Logger("Doc ~s field ~s is not a string: ~p", [Id, Field, Non_bin]),
                    false
                ; Value ->
                    %Logger("Testing '~s' to '~s'", [Value, Regex]),
                    case re:run(Value, Regex, [])
                        of nomatch ->
                            %Logger("No match: ~s", [Id]),
                            false
                        ; _ ->
                            %Logger("Match: ~s", [Id]),
                            true
                        end
                end

            % Matches of course pass. But if the match fails and ddocs are requested, pass any ddocs.
            , case Is_match
                of true -> true
                ;  false
                    -> Id = couch_util:get_value(<<"_id">>, Doc)
                    , DDocs = couch_util:get_value(<<"ddocs">>, Query)
                    , Is_ddocs = DDocs =:= true orelse DDocs =:= <<"true">>
                    , case {Is_ddocs, Id}
                        of {true, <<"_design/", _/binary>>} -> true
                        ;  _                                -> false
                        end
                end
        end
    end
.

% vim: sts=4 sw=4 et

File: js/filters/refilter.js
// Typical req.query (at a minimum, the "filter" parameter will be there):
//
// { "filter": "couchdb/refilter"
// , "field" : "_id"
// , "regex" : "^a"
// , "log"   : true
// }

function(doc, req) {
  var field = req.query.field
    , regex = req.query.regex
    , is_ddocs = (req.query.ddocs == 'true' || req.query.ddocs == true)

  logger('Running: id=%s req=%o', doc._id, req)

  if(is_ddocs && doc._id.match(/^_design\//)) {
    logger('Passing design doc: %s', doc._id)
    return true
  }

  // It seems like throw causes a 500 in CouchDB.
  if(typeof field != 'string') {
    //throw new Error('Required "field" string parameter')
    log('ERROR: invalid_query: Required "field" string parameter')
    return false
  }

  if(typeof regex != 'string') {
    //throw new Error('Required "regex" string parameter')
    log('ERROR: invalid_query: Required "regex" string parameter')
    return false
  }

  logger('field=%s regex=%s', field, regex)

  var value = doc[field]

  if(typeof value == 'undefined') {
    logger('Doc %s has no field %s', doc._id, field)
    return false
  }

  if(typeof value != 'string') {
    logger('Doc %s field %s is not a string: %o', doc._id, field, value)
    return false
  }

  logger("Testing '%s' to '%s'", value, regex)
  regex = new RegExp(regex)
  if(! value.match(regex)) {
    logger('No match: %s', doc._id)
    return false
  } else {
    logger('Match: %s', doc._id)
    return true
  }

  function logger(message) {
    if(!req.query.log)
      return

    var params = Array.prototype.slice.apply(arguments, [1])
    if(params.length > 0)
      message += ' <- ' + JSON.stringify(params)

    log(message)
  }
}

// vim: sts=2 sw=2 et
